File: Java/387-FirstUniqueCharacterinaString.java
//two pass
public class Solution {
    public int firstUniqChar(String s) {
        int[] dict = new int[26];
        for (int i = 0; i < s.length(); i++){
            dict[s.charAt(i) - 'a']++;
        }
        for (int i = 0; i < s.length(); i++){
            if (dict[s.charAt(i) - 'a'] == 1)
                return i;
        }
        return -1;
    }
}

//one pass
//Actually one string scan + one frequency array scan.
public class Solution {
    public int firstUniqChar(String s) {
        char[] schar = s.toCharArray();
        int result = schar.length;
        int[] seen = new int[26];
        //-1 => not visited
        //-2 => repeating
        //>=0 => appear once, value is index
        Arrays.fill(seen, -1);
        for(int i = 0; i < schar.length; i++) {
            int index = schar[i]-'a';
            if(seen[index] == -1) seen[index] = i;
            else if(seen[index] >= 0) seen[index] = -2;
        }
        for(int i = 0; i < 26; i++) {
            if(seen[i] >= 0) result = Math.min(result, seen[i]);
        }
        return result == schar.length?-1:result;
    }
}

File: cpp/387-FirstUniqueCharacterInAString.cpp
class Solution {
public:
    int firstUniqChar(string s) {
        //-2: duplicated; -1: initial val, since index might be 0
        int count[26] = {};
        for (int i = 0; i < 26; i++)
            count[i] = -1;
        for (int i = 0; i < s.size(); i++) {
            if (count[s[i] - 'a'] == -1) {
                count[s[i] - 'a'] = i;
            } else if (count[s[i] - 'a'] >= 0) {
                count[s[i] - 'a'] = -2;
            }
        }
        int first = INT_MAX;
        for (int c : count) {
            if (c >= 0)
                first = min(first, c);
        }
        return first == INT_MAX ? -1 : first;
    }
};
